File: ai/src/time_curves.py
"""
LoL Draft AI - Time-Horizon Win-Rate Curves (Python AI Mirror)
Predicts early, mid, and late game win rates and generates smooth time-horizon trajectories.
"""

import math
from typing import List, Dict, Any, Optional
from eval_bar import winrate_to_eval_bar, DEFAULT_SCALE

TIME_INTERVALS = [10, 15, 20, 25, 30, 35, 40]


def resolve_phase(minute: int) -> str:
    if minute <= 15:
        return "EARLY_GAME"
    elif minute <= 25:
        return "MID_GAME"
    elif minute <= 35:
        return "LATE_GAME"
    else:
        return "ULTRA_LATE"


def generate_trajectory_summary(early_wr: float, late_wr: float) -> str:
    if early_wr >= 0.52 and late_wr <= 0.48:
        return "Early Bully -> Late Falloff"
    elif early_wr <= 0.48 and late_wr >= 0.52:
        return "Early Deficit -> Late Scaling Inversion"
    elif early_wr >= 0.52 and late_wr >= 0.52:
        return "Sustained Blue Composition Dominance"
    elif early_wr <= 0.48 and late_wr <= 0.48:
        return "Sustained Red Composition Dominance"
    else:
        return "Evenly Contested Trajectory"


def predict_time_curve(
    baseline_blue_win_rate: float,
    laning_delta: float = 0.0,
    late_scaling_delta: float = 0.0,
    early_bully_delta: int = 0,
    hyper_carry_delta: int = 0,
    early_dominance_delta: float = 0.0,
    matchup_delta: float = 0.0,
    durability_delta: float = 0.0,
    cc_delta: float = 0.0,
    scale: float = DEFAULT_SCALE,
) -> Dict[str, Any]:
    safe_base = max(0.01, min(0.99, baseline_blue_win_rate))
    base_logit = math.log(safe_base / (1.0 - safe_base))

    early_signal = (
        laning_delta * 0.12
        + early_bully_delta * 0.10
        + early_dominance_delta * 0.04
        + matchup_delta * 0.08
    )
    late_signal = (
        late_scaling_delta * 0.12
        + hyper_carry_delta * 0.10
        + durability_delta * 0.03
        + cc_delta * 0.03
    )

    points = []
    for minute in TIME_INTERVALS:
        w_early = 1.0 / (1.0 + math.exp((minute - 16.0) / 4.0))
        w_late = 1.0 / (1.0 + math.exp(-(minute - 27.0) / 4.5))

        shift = early_signal * (w_early - 0.35) + late_signal * (w_late - 0.35)
        logit = base_logit + shift
        blue_wr = round(max(0.01, min(0.99, 1.0 / (1.0 + math.exp(-logit)))), 4)
        red_wr = round(1.0 - blue_wr, 4)

        eval_bar = winrate_to_eval_bar(blue_wr, scale=scale)
        phase = resolve_phase(minute)

        points.append({
            "minute": minute,
            "blue_win_rate": blue_wr,
            "red_win_rate": red_wr,
            "eval_bar": eval_bar,
            "dominant_phase": phase,
        })

    min10 = next(p["blue_win_rate"] for p in points if p["minute"] == 10)
    min15 = next(p["blue_win_rate"] for p in points if p["minute"] == 15)
    min20 = next(p["blue_win_rate"] for p in points if p["minute"] == 20)
    min25 = next(p["blue_win_rate"] for p in points if p["minute"] == 25)
    min30 = next(p["blue_win_rate"] for p in points if p["minute"] == 30)
    min35 = next(p["blue_win_rate"] for p in points if p["minute"] == 35)
    min40 = next(p["blue_win_rate"] for p in points if p["minute"] == 40)

    early_wr = round((min10 + min15) / 2.0, 4)
    mid_wr = round((min20 + min25) / 2.0, 4)
    late_wr = round((min30 + min35) / 2.0, 4)
    ultra_late_wr = round(min40, 4)

    summary = generate_trajectory_summary(early_wr, late_wr)

    return {
        "points": points,
        "early_game_win_rate": early_wr,
        "mid_game_win_rate": mid_wr,
        "late_game_win_rate": late_wr,
        "ultra_late_win_rate": ultra_late_wr,
        "trajectory_summary": summary,
    }

File: src/main/kotlin/com/loldraft/models/TimeCurves.kt
package com.loldraft.models

import com.loldraft.data.meta.ChampionTag
import com.loldraft.data.meta.ChampionTagRegistry
import com.loldraft.data.meta.PatchMetaMatrix
import com.loldraft.data.meta.PowerSpikeCurve
import com.loldraft.data.models.DraftState
import com.loldraft.data.style.TeamTacticalProfile
import kotlinx.serialization.Serializable
import kotlin.math.exp
import kotlin.math.ln
import kotlin.math.round

@Serializable
enum class MatchPhase {
    EARLY_GAME,
    MID_GAME,
    LATE_GAME,
    ULTRA_LATE,
}

@Serializable
data class TimePointWinRate(
    val minute: Int,
    val blueWinRate: Double,
    val redWinRate: Double,
    val evalBar: EvalBarScore,
    val dominantPhase: MatchPhase,
    val keyFactors: List<String> = emptyList(),
)

@Serializable
data class TimeCurve(
    val points: List<TimePointWinRate>,
    val earlyGameWinRate: Double,
    val midGameWinRate: Double,
    val lateGameWinRate: Double,
    val ultraLateWinRate: Double,
    val trajectorySummary: String,
)

class TimeCurveCalculator(
    val tagRegistry: ChampionTagRegistry = ChampionTagRegistry.createDefault(),
    val evalBarScale: Double = EvalBarCalculator.DEFAULT_SCALE,
    val earlyInflectionMinute: Double = 15.5,
    val lateInflectionMinute: Double = 28.5,
    val earlyInflectionSlope: Double = 4.0,
    val lateInflectionSlope: Double = 4.5,
    val weightLaning: Double = 0.08,
    val weightEarlyBully: Double = 0.06,
    val weightDominance: Double = 0.04,
    val weightMatchup: Double = 0.08,
    val weightLateScaling: Double = 0.08,
    val weightHyperCarry: Double = 0.06,
    val weightDurability: Double = 0.02,
    val weightCc: Double = 0.02,
) {
    companion object {
        val TIME_INTERVALS: List<Int> = listOf(10, 15, 20, 25, 30, 35, 40)
        private const val EPSILON: Double = 1e-6
    }

    fun calculate(
        draftState: DraftState,
        features: DraftFeatures,
        baselineBlueWinRate: Double,
        patchMeta: PatchMetaMatrix? = null,
        blueTeamProfile: TeamTacticalProfile? = null,
        redTeamProfile: TeamTacticalProfile? = null,
    ): TimeCurve {
        val safeBaseP = baselineBlueWinRate.coerceIn(0.01, 0.99)
        val baseLogit = ln(safeBaseP / (1.0 - safeBaseP))

        val blueProfiles = draftState.bluePicks.mapNotNull { tagRegistry.getProfile(it.championId) }
        val redProfiles = draftState.redPicks.mapNotNull { tagRegistry.getProfile(it.championId) }

        val blueEarlyCount =
            blueProfiles.count {
                it.powerSpike == PowerSpikeCurve.EARLY_SPIKE || it.tags.contains(ChampionTag.EARLY_BULLY)
            }
        val redEarlyCount =
            redProfiles.count {
                it.powerSpike == PowerSpikeCurve.EARLY_SPIKE || it.tags.contains(ChampionTag.EARLY_BULLY)
            }
        val deltaEarlyBully = blueEarlyCount - redEarlyCount

        val blueLateCount =
            blueProfiles.count {
                it.powerSpike == PowerSpikeCurve.HYPER_SCALING ||
                    it.powerSpike == PowerSpikeCurve.LATE_GAME_SPIKE ||
                    it.tags.contains(ChampionTag.HYPER_CARRY)
            }
        val redLateCount =
            redProfiles.count {
                it.powerSpike == PowerSpikeCurve.HYPER_SCALING ||
                    it.powerSpike == PowerSpikeCurve.LATE_GAME_SPIKE ||
                    it.tags.contains(ChampionTag.HYPER_CARRY)
            }
        val deltaHyperCarry = blueLateCount - redLateCount

        val deltaLaning = features.radarDelta.laningStrength
        val deltaLateScaling = features.radarDelta.lateGameScaling
        val deltaDominance = features.earlyDominanceDelta
        val deltaMatchup = features.matchupDelta
        val deltaDurability = features.durabilityDelta
        val deltaCc = features.ccDelta

        val earlySignal =
            deltaLaning * weightLaning +
                deltaEarlyBully * weightEarlyBully +
                deltaDominance * weightDominance +
                deltaMatchup * weightMatchup

        val lateSignal =
            deltaLateScaling * weightLateScaling +
                deltaHyperCarry * weightHyperCarry +
                deltaDurability * weightDurability +
                deltaCc * weightCc

        val points = mutableListOf<TimePointWinRate>()

        for (minute in TIME_INTERVALS) {
            val wEarly = 1.0 / (1.0 + exp((minute - earlyInflectionMinute) / earlyInflectionSlope))
            val wLate = 1.0 / (1.0 + exp(-(minute - lateInflectionMinute) / lateInflectionSlope))

            val shift = earlySignal * (wEarly - 0.35) + lateSignal * (wLate - 0.35)
            val logit = baseLogit + shift
            val rawBlue = 1.0 / (1.0 + exp(-logit))
            val blueWinRate = roundToFourDecimals(rawBlue.coerceIn(0.01, 0.99))
            val redWinRate = roundToFourDecimals(1.0 - blueWinRate)

            val evalBar = EvalBarCalculator.calculate(blueWinRate, scale = evalBarScale)
            val phase = resolveMatchPhase(minute)
            val factors = extractKeyFactors(minute, phase, deltaLaning, deltaEarlyBully, deltaLateScaling, deltaHyperCarry)

            points.add(
                TimePointWinRate(
                    minute = minute,
                    blueWinRate = blueWinRate,
                    redWinRate = redWinRate,
                    evalBar = evalBar,
                    dominantPhase = phase,
                    keyFactors = factors,
                ),
            )
        }

        val min10 = points.find { it.minute == 10 }?.blueWinRate ?: safeBaseP
        val min15 = points.find { it.minute == 15 }?.blueWinRate ?: safeBaseP
        val min20 = points.find { it.minute == 20 }?.blueWinRate ?: safeBaseP
        val min25 = points.find { it.minute == 25 }?.blueWinRate ?: safeBaseP
        val min30 = points.find { it.minute == 30 }?.blueWinRate ?: safeBaseP
        val min35 = points.find { it.minute == 35 }?.blueWinRate ?: safeBaseP
        val min40 = points.find { it.minute == 40 }?.blueWinRate ?: safeBaseP

        val earlyGameWinRate = roundToFourDecimals((min10 + min15) / 2.0)
        val midGameWinRate = roundToFourDecimals((min20 + min25) / 2.0)
        val lateGameWinRate = roundToFourDecimals((min30 + min35) / 2.0)
        val ultraLateWinRate = roundToFourDecimals(min40)

        val summary = generateTrajectorySummary(earlyGameWinRate, lateGameWinRate)

        return TimeCurve(
            points = points,
            earlyGameWinRate = earlyGameWinRate,
            midGameWinRate = midGameWinRate,
            lateGameWinRate = lateGameWinRate,
            ultraLateWinRate = ultraLateWinRate,
            trajectorySummary = summary,
        )
    }

    private fun resolveMatchPhase(minute: Int): MatchPhase =
        when {
            minute <= 15 -> MatchPhase.EARLY_GAME
            minute <= 25 -> MatchPhase.MID_GAME
            minute <= 35 -> MatchPhase.LATE_GAME
            else -> MatchPhase.ULTRA_LATE
        }

    private fun extractKeyFactors(
        minute: Int,
        phase: MatchPhase,
        laningDelta: Double,
        earlyBullyDelta: Int,
        lateScalingDelta: Double,
        hyperCarryDelta: Int,
    ): List<String> {
        val factors = mutableListOf<String>()
        when (phase) {
            MatchPhase.EARLY_GAME -> {
                if (laningDelta > 0.5) factors.add("Blue laning strength priority")
                if (laningDelta < -0.5) factors.add("Red laning strength priority")
                if (earlyBullyDelta > 0) factors.add("Blue early bully spike aggression")
                if (earlyBullyDelta < 0) factors.add("Red early bully spike aggression")
                if (factors.isEmpty()) factors.add("Evenly matched laning phase")
            }
            MatchPhase.MID_GAME -> {
                factors.add("Objective contest and tower transition")
            }
            MatchPhase.LATE_GAME, MatchPhase.ULTRA_LATE -> {
                if (lateScalingDelta > 0.5) factors.add("Blue hyper-scaling teamfight advantage")
                if (lateScalingDelta < -0.5) factors.add("Red hyper-scaling teamfight advantage")
                if (hyperCarryDelta > 0) factors.add("Blue carries item spikes reached")
                if (hyperCarryDelta < 0) factors.add("Red carries item spikes reached")
                if (factors.isEmpty()) factors.add("Late game 5v5 teamfight execution")
            }
        }
        return factors
    }

    private fun generateTrajectorySummary(
        earlyWinRate: Double,
        lateWinRate: Double,
    ): String =
        when {
            earlyWinRate >= 0.52 && lateWinRate <= 0.48 -> "Early Bully -> Late Falloff"
            earlyWinRate <= 0.48 && lateWinRate >= 0.52 -> "Early Deficit -> Late Scaling Inversion"
            earlyWinRate >= 0.52 && lateWinRate >= 0.52 -> "Sustained Blue Composition Dominance"
            earlyWinRate <= 0.48 && lateWinRate <= 0.48 -> "Sustained Red Composition Dominance"
            else -> "Evenly Contested Trajectory"
        }

    private fun roundToFourDecimals(value: Double): Double = round(value * 10000.0) / 10000.0
}
